### ISSUE TITLE
Notepad++ is corrupting files and opening them with null values.

### ISSUE THREAD

**Issue by nightshaddow13** [notepad-plus-plus/notepad-plus-plus#4655]
### Description of the Issue
Occasionally when my computer goes to sleep and I have a '.pl' file open in notepad++ the file will become corrupt.  This has happened too me two times in the last 2 days.  The text seems to be completely deleted.

### Steps to Reproduce the Issue
1. Have open perl script
2. Put computer to sleep
3. Wake up device and open the document

### Expected Behavior
I expected my data to remain unscathed.

### Actual Behavior
The data was replaced with null data

### Debug Information
Notepad++ v7.5.7   (64-bit)
Build time : Jun 29 2018 - 00:53:58
Path : C:\Program Files\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS : Windows 10 (64-bit)
Plugins : DSpellCheck.dll mimeTools.dll NppConverter.dll 

![image](https://user-images.githubusercontent.com/41086688/42536703-c7bbb80e-8460-11e8-96f6-717bc9db7838.png)

**Comment by sasumner** [comment]
Duplicate of #2381 and perhaps others.  Also see https://notepad-plus-plus.org/community/topic/15748/faq-desk-crash-caused-my-file-to-be-all-nulls

Long-standing, known issue...sad.  The #1 thing a text editor absolutely has to do is to not corrupt and not lose data.

**Comment by sasumner** [comment]
@nightshaddow13 , has this problem recurred for you?  We're looking for a reliable reproduction sandbox.

**Comment by pryrt** [comment]
@nightshaddow13, you seem to be the first I've seen here or in the [Community Forums](https://notepad-plus-plus.org/community/) that claims a reliable / repeatable method for causing the NULs (most people have it happen once, and never see it again).  How willing are you to help with the debug?

I don't have a build environment that would work for NPP, but there are plenty here who do.  I wonder if someone could create a debug-build executable for you, which would log extra debug info to a log file, flushing often: dump of all the settings (in case it's snapshot on/off, remember session on/off, etc), as well as recording various internal function calls -- if developers had a trace of the last few functions (or the last function that had enough time to flush the logfile to disk) that ran before the crash event, it might go a long way toward finding (and solving) the problem.

**Comment by nightshaddow13** [comment]
I would be more than willing to help.  I will try a few things to see if I can get it to happen again.  I recently reinstalled windows thinking it would help with that and maybe some other issues but it has not.  I have been forced to backup my files regularly to prevent a loss of data.

**Comment by pnedev** [comment]
Hello @nightshaddow13 ,

Was the backup N++ functionality enabled when the issue happened?
Thank you.

**Comment by prince2007g** [comment]
E:\FRIENDS-PROJECT\src\styles\sass
website.scss is corrupted due after restarting the system, the file showing the same size as it should be but file content showing NULL, null, I did not find in appdata/Roam/Notepad++/backup
Date: 15/12/2018 time around 9:48am
Please suggest how to recover its very important, the complete project on it.

**Comment by sasumner** [comment]
@prince2007g A very important detail you left out:  Which version of Notepad++ are you running?

**Comment by prince2007g** [comment]
i think the version was version 6.9.1, i am not sure. Now the version is 7.6.1.
I restarted the system that time website.scss was opened in notepade++, After restarted the system i saw Notepad also updated and file website.scss converted with encoded content NULL, NULL.....

**Comment by pnedev** [comment]
@prince2007g ,

I couldn't quite understand your answer.
The version where the problem occurred was 6.9.1, correct?

Unfortunately there is not much you can do.
In the forums people suggested using Recuva (or other program fro restoring accidentally deleted files).
You may try using Windows _'Previous Version'_ function - as far as I remember it is accessible through Windows file explorer - right-click on the file and there (or in the file Properties) there should be option like _'Previous versions'_.
The last suggestion from a user in the forum was to open the file in a browser (how can one do that?!). The user claimed it was opened with it's original content (not all NUL) but I don't know details and it doesn't make sense to me at all.

**Comment by BugZappr** [comment]
I'm getting the same stuff on vers. 7.8.5.  I love NPP, but I don't have time to come up with a test case,.  It rarely happens., and is hard to catch.  Unfortunately, this was a password file; so I won't be using NPP in the foreseeable future, that is, until this is found. 

I can tell you that I don't set my computer to suspend or to hibernate; but I do at times, and probably in this case, shut down the computer with NPP open.  Sometimes, and possibly in this case, it is with an unsaved file.  It always seemed to remember them, and let me save them when I restarted NPP.

I had Autosave installed and set to 10 min.; but oddly enough in this case, I found it not activated.  This had happened before, with another file that I thought was important; but I thought it might have been my disk encryption software or a hacker.  Now, NPP has been caught doing the same thing, with the same NULLs where my password file should be.

Now, to try to find out where my Stimulus check is having just lost all the login info to the Intuit site the IRS uses.  Fun stuff.

**Comment by swift1978** [comment]
At the Moment my blood is boiling. I lost 3 days worth of hard work on a Saved file, because of an Unexpected Power outage although I was on UPS backup. And I open the file to find all NULL characters. I have been frantically trying to recover a previous version, and rampaging around googling forums, and tips. Just tried Recuva, with different settings, but it is not able to recover, since the file has been over-written. Have given up that what is gone is gone. Notepad++ has devoured 3 days of prime time in my life, and I have to do everything all over again, if that is even possible.

I am surprised that this problem has been active on this thread since 16-Feb-2017, and who knows if it was even earlier.

Notepad++ Developers, Contributors, Volunteers, yes, your Program is a Fantastic program and now probably very widely used throughout the world.

BUT GET THIS INTO YOUR HEADS, THAT WHAT IS MOST IMPORTANT, MORE IMPORTANT THAN ANY FEATURES AND FRILLS YOU PROVIDE, IS ‘DATA INTEGRITY’, ‘STABILITY’, ‘RELIABILITY’, DEPENDABILITY. IF YOU DON’T PROVIDE WHATEVER FEATURE AND FRILL YOU PROVIDE ON THE BACK OF THESE FUNDAMENTALS, THEN ALL YOU GIVE IS ONLY HOGWASH, AND YOU WILL ONLY HAVE USERS CURSING YOU, AND WISHING THEY HAD NEVER USED YOUR PROGRAM EVER AT ALL.

Can’t you even check what other Editors do. Even MsWord, Excel and the like are able to diagnose Data Corruption, and offer an AutoRecover feature. With Notepad++, even with Backups on, and every setting you can find ‘on’, and with Regular saving of the File, still we are stuck with Null Files.

Is this some Advanced Coding Logic?? To me, this is something even a Numbskull would think of.

When you open a file, can you not program some logic, and a list of parameters that would check the file versions opened for every kind of corruption in our human experience, and confirm file integrity first.
Especially when the previous Closing of Notepad has not been ‘normal’.
If you sense File Corruption, Session Corruption why don’t you alert the User, and have some Options ready for them to check and recover, or have some algorithms by which you can reasonably well confirm the best previous reliable state. Before you go on this filling with Null Characters over-drive, and destroying files and data. Is it because you don’t care because it is free, open-source, and people have to just bow to you and keep thanking you because it is free…

**Comment by sasumner** [comment]
@swift1978 Supposedly the problem if fixed with the recent https://github.com/notepad-plus-plus/notepad-plus-plus/commit/a1031517742c7ef600b825bf9f193b3f2740dff6

**Comment by swift1978** [comment]
Hope so. I am still tingling all over, trying to cope with and rebuild all
that data I lost...

**Comment by swift1978** [comment]
It would help no one, neither me, you, nor us, if we choose to be Evasive, and take Comfort in passing the Buck to Microsoft. Regardless of the OS, these facts still stand out, and these are the Questions we need to ask ourselves.
Leaving the OS aside, what is the Commitment to Data Integrity Strategy that NPP applies.
1) Does Notepad, every time it launches, have a check, a marker, where it evaluates the last shutdown, and only when it confirms that everything is kosher, then proceed?
2) Does it Maintain a Marker in the First place? Not too tough to put that in place. (I am not a Coder, I am an Engineer and Manager and all I can offer is Common-sense and simple logic). One can have a Mirror Save of all the Edit (Undo) information that each file has. (Which is in Memory, which kicks in when we press Undo on each file). Plus the File Content, actually, whatever is in Memory. A Mirror could be maintained, a locked one for NPP alone, and unreadable for other programs. Not many copies of it but the Current. 
	(Guess this may not be too important, though: And this Mirroring Logic also checks, if the current one that it is using to over-write is immaculate, and whatever changes are coming are the usual human changes (an algorithm could check that). So that the result of some crash, or misadventure, is not used to over-write authentic data...)
In any case, with this Strategy, I have an incorrupted state, always being maintained, one copy of it, saved somewhere in my hard disk and not purely in Memory.
3) And now, I come to two Scenarios. Have a Regular PC shutdown, a regular Closing of NPP (or) some System Crash, or Blue Screen, or whatever. NPP registers in the Marker, the Type of the Shutdown, whether it is Kosher and beyond any doubt (or) whether it is a Crash
4) When the NPP is relaunched, if it is Kosher, it restarts the usual way. (I would like the Undo Information also to be reloaded, but if that would make the Memory demand too high, the User could be given the option to refresh the file, and take off all the Undo information, if he is totally assured about the Present State for each file) 
5) But if THE LAST SHUTDOWN WAS A CRASH, it checks and uses some internal Compare facility, and checks on various parameters, the RAM version against the Saved Marker, and in whichever file there is a difference, it brings it up on Screen. That way the user is not taken by surprise, and left bewildered, but has all the options on all that is possible, and can take the best decisions to preserve his data.
(WHATEVER HAPPENS WITH THE OS, NPP DOES NOT & SHOULD NOT LOAD A NULL FILE. I CAN'T UNDERSTAND THE LOGIC ON WHY IT DOES THAT AT PRESENT. WHO WOULD EVER AUTHORIZE THE LOADING OF A NULL FILE! OF WHAT USE IS A NULL FILE TO ANY ONE! LOADING THAT IS MISCHIEVOUS! YOU COULD RATHER ALERT THE USER THAT IT IS NULL. WHICH YOU COULD IF YOU HAD A CHECK ROUTINE THAT CHECKS FOR DATA INTEGRITY). Ok, you might blame that Microsoft was the Idiot who thought it fit to load that Null Character file into Memory. But is NPP also not an Idiot to think it fit to load that Null Character file into it's Memory and display it on Screen. For any Auto-save plugins to then save that Null file over the previously Saved files and wreak havoc. 
The Whole Set of Code that allows the Null file to be loaded and saved is mischievous. Makes me kind of agree with Ben's Sarcastic renaming of Notepad as Nullpad.

**Comment by swift1978** [comment]
Yes while having an OS-independent Strategy like above, let's also in addition to that, make full use of the OS' provisions at the same time. It is definitely wise and of value to check with the OS on 'how' the OS dumps files in their latest updated avatar. And properly understand how they are doing it, the latest parameters for those commands. And certainly test whatever you have coded, and then check that the results are as you intended. And if not pursue that and fix it. If the OS requires a 'commit', it is their OS, and they have the right to modify their code. Our part should have been to TRY OUT, and TEST. If we had done that, we would have found EARLY that there is a problem with the way the OS is handling the Dump. And that the command is not yielding the expected results. A little pursuing would have informed us much early that a commit is required, and not had us leaving a trail of destruction behind us.

**Comment by swift1978** [comment]
I am Repeating myself, but really can't understand why NPP doesn't keep a (Internal Mirror Save) of the File State, and Edit/Undo Information that each File has, which at the Moment they have in Memory; Written and Saved into the Hard-Disk as a System File, that will be used if there is any Corruption or Deletion in the Undo Information in Memory.
The Reason I say this is that THERE IS NOTHING MORE DESERVING than Notepad of my System Resources, Hard Disk space and Memory. And I believe Everyone would agree. It's not like Text Files take that much space, either. Much of every other thing we do on the PC, comes in the Recreational Category. NPP is where we bring out our Imagination, and where our Real Work gets Captured, especially for Coders and Writers, it is our Juice. And for many of us, that probably is the Main Reason we have our PC for. Atleast for me, I would gladly commit to NPP as much resources as it wants. NPP is critically important.

**Comment by swift1978** [comment]
Data Losses make us Desperate, and often to go into over-reacting responses which cause damage of their own. And since NPP throws shit on our faces once in a while, we have to browse, search, learn/plan/include all kinds of Strategies, making it a part of how we work. We shouldn't need to, if NPP would take Data Integrity seriously. We do all this Frantic pressing of Save, and Save all, all the time, because you never know when it would crash, and destroy all it was keeping in RAM memory. We get Auto-Save plugins and all. Would we need to do that, if the Fundamental aspect of Data Integrity was kept close by the NPP team. And then the Auto Save that was to help us not lose data over-writes Good Data with what NPP loaded into Memory, which was Clearly recognizable Bad Accident, a Null File.

**Comment by swift1978** [comment]
ABOUT NPP'S SESSION MANAGEMENT STRATEGY, AND HOW MANY HOLES WE HAVE HERE TOO...
1) In the Backup Options Window in NPP, we have an option to enable Session Snapshot and PERIODIC Backup. The Impression is that there would be a Periodic Backup of Dated/Timed Sessions, possibly with a NPP Determined on the Number of Sessions it would retain, before starting to over-write the Oldest. They have a Backup Path. But the Fact is that No Session Information gets saved there. That word BACKUP makes no sense there. What on earth are you backing up and where, when the User selects that option. You could check in your NPP. So, all that Option and information is all highly misleading, isn't it. Again shows very poor commitment to Data Integrity.

2) Then you have a Backup on Save Option. Which I am sure most would not be able to make a Sense of whatever that could mean. If I am saving, I am already saving a file, and then what's the intention of the Backup at that time. And NPP also gives two options there, Simple Backup, or Verbose Backup. Should some kind of a clarifying tool tip not be provided in those options. Especially when the logic of these options is something strange, and people would wonder whatever is this going to do. If this was a Bells & Whitles option, sure, we could just try it out, take potshots, and find when it hits. But this is about Data Integrity, and we can't take a chance here. We should know for sure, what that option would do, and then reliably use it to get definitive results. Tool Tip explaining the Logic and the Thought Flow of these Options is vital, what it would do, so we know what it does not.

3) A Good Program should not only preserve Data, but also the State, the Position of each file, and also include any Edit/Undo Data that the User could use to Rollback/Undo.
So, we ensure that all possible Scenarios in which each of these could come under attack. A Couple that I can think of is when a File that is open in NPP, is also opened in another program, edited and saved there. Or if a Folder or File move is done, and the File Location changed. A Software techie could think of many more scenarios.
But how does NPP function in these. Do a Folder Move, when NPP is closed, and it removes the moved files from the Session, just deletes them and over-writes the Session. It does not give the User any information that it is not able to find these files, should they point to another location, have the files been moved, do you want to retain that file data in memory, and save a copy, or save it over the moved file. Nothing. It's like NPP is so trigger-happy to delete and over-write all the Session Info. And since there is no Periodic Backup, there is no option to go back to the Previous Session, and have these files re-opened. NPP is just on a Deleting, Over-writing Overdrive, leaving the User High and Dry. It is unconscionable.
A well-designed program should make the user relaxed and care-free, and feel totally assured about data integrity, and not on the tenterhooks snorting fire as Notepad++ often gets us to.

**Comment by swift1978** [comment]
What NPP shouldn't become: Develop an apathy that has it's roots somewhere in the thought that what we are doing is charity, and you are anyway not deserving of whatever you are getting. Did you pay anything for it, then how dare you complain. Take it or leave it, and that 'who cares' kind of an attitude. That attitude displays no pride about the software, no quest for excellence. No sense of just how Huge Notepad++ is, and how many throughout the whole world depend on it.

NPP approach cannot be. Let someone die first, and then if their relatives raise an issue with us, and we get the bug, only then we'll raise a ticket, and then look at fixing it when we can spare the time. If something isn't broken, and is somehow moving along, don't touch it attitude.
That is not how you get to be a Flagbearer.
Data Integrity Strategy, should be system independent, and it should be periodically reviewed, and ensured that it is working in a nuclear, self-contained, and reliable manner. This is not a feature/frill issue that you tackle only when there is a bug. This is fundamental, and foundational. It has to be perfect, all the time.

Notepad++ had already got me to tenterhooks, and even before this Null file issue, I was nearly paranoid in clicking Save all, periodically. There is a time in which there is no issue, and just when you are getting a little comfortable that your Save Strategy is ensuring some integrity, Notepad++ drops another bombshell, and boom goes another set of data. And when you seek assistance, there are ever so many people asking you to make solutions of your own, go to the cloud, use this backup that backup. Don't expect us to care about your Data integrity. All we give is just a Toy to play with, not something you can rely on. That was never our intention. Don't you'll have any shame.
It's ok if someone really doesn't care about their data. But where it is invaluable, and people hunt for methods, strategies, by which they can avoid data loss, there should atleast be some attention from the developers to provide options by which users can ensure that. Actually, by default, data safety options should be automatically enabled, more than leaving it to the user's choice.

**Comment by pnedev** [comment]
@swift1978 ,

User data (files) ultimately lives in the file system (that is the operating system itself).
So the program's responsibility is to instruct the Operating System to do the saving of the user data BUT it is the Operating System itself that provides that functionality and ultimately performs the saving.

After commit https://github.com/notepad-plus-plus/notepad-plus-plus/commit/a1031517742c7ef600b825bf9f193b3f2740dff6 Notepad++ will instruct the OS to flush all its write buffers and cache to disk so neither Notepad++ crash nor OS crash (or sudden power loss) will lead to data loss.

Notepad++ itself has backup setting with periodic background savings and makes additional copies of unsaved files in its own folder that have timestamp and are offered for restore when you reload your session (re-open your files that didn't save properly).

The backups were not properly saved by the OS as well before the https://github.com/notepad-plus-plus/notepad-plus-plus/commit/a1031517742c7ef600b825bf9f193b3f2740dff6 commit so they were not valid when power loss happened.

In Notepad++ v7.9.1 (that will include the fix) the user data should be properly secured.

**Comment by sasumner** [comment]
>The backups were not properly saved by the OS as well before the a103151 commit so they were not valid when power loss happened.

I can fully confirm this, well, not with a power loss but with a BSOD.
When I restarted and was shocked to see my files saying NUL NUL NUL, I thought, no matter, I'll just go get the backups N++ made.
My shock was total and complete when I saw the backups were NUL NUL NUL as well.

**Comment by BugZappr** [comment]
Does it bother anybody else how quickly this issue went from "Supposedly the problem if fixed" to closed?  This is the sort of issue that makes N++ less than useless if it is still present.  I hope it has fixed it.  I liked N++ very well for many years, once it started underlining spelling errors in red (which I posted an enhancement req. for).  I look forward to returning to it.  This problem was/is so severe, that I've never encountered a text editor nuking files like this.  It happened to me 3X, on files I cared about, and it had been at least some time since I'd backed each up in these cases alone.  I will keep checking here to see if this horrible issue is really fixed; but I won't start using it again for perhaps several months, to be sure.  I'm limping along with Kate in Linux and jEdit, right now.

Everybody, please post any recurrences of this problem here.  I'll be checking back, to see if it is fixed.  IMHO, this bug should have never have been closed until some months had gone by, to be sure it is fixed.  It seems a really bad energy if it is only "Supposedly the problem if fixed"; when it is so severe, and not easily testable; to close it so quickly without really knowing.

To look at the commit, it seems ridiculous that M$ would turn on write-behind caching by default, playing fast and dangerous with user data for the sake of benchmarks.  Then, when you call fflush(), the M$ description SAYS it writes the data out to the file; but from the commit, the generic_fopen() (whatever that is), must have the 'c' as part of its mode, to actually write out the data.  The fflush() alone should have done the job.  It is understandable that a programmer would have every right to expect that the data had been written to disk at that point.  To those of us who aren't chumps, we've turned off "write behind caching" on every system we've installed.  Thus, it isn't surprising it remained unnoticed.  Yes, this seems to be entirely a product of insipid M$ reasoning.

That being said, it seems in general that closing issues just because a patch has been committed that SUPPOSEDLY fixes it is just bad practice and "jumping the gun."

**Comment by swift1978** [comment]
Well, I don't get this. Backups from any Program are done by that Program
itself - definitely not the OS. Any NPP Backups would be done by NPP, and
NPP would be responsible for what it Backed up.

Power Loss, or BSOD you saw your files saying Nul, Nul, that was because
NPP rebooted from it's State in Memory, and did not open the File on Disk.
Memory Handling is OS' Jurisdiction. OS didn't handle it, dump it properly,
or NPP didn't instruct the OS well on the way to Commit it, whatever. What
was in Memory was corrupted, probably not the File on Disk.

 The Problem with your File on Disk getting Corrupted, was that NPP not
only loaded a Null File on Screen, and into Memory after Re-Launch, but
further compounded it by saving it to Disk, and over-writing the Intact
File, losing it for all time, in one act. It may not be NPP by itself as
such; it could be some Plugin that Saves the Files automatically... Since
the Undo/Edit information does not get loaded on Re-Launch of NPP, you
cannot Undo it back to what it was. The Over-write is Final and
Irredeemable. The Old File is done with, and all you have is a File full of
precious Nuls.

And to add to having blown away the Legit File with a File Full of Nuls,
NPP would not have stopped there. We all are so anxious about our Data that
we have this very frequent Save and Backup options enabled. So, NPP would
have gone ahead, and even over-written the Backup with the Nul file,
ensuring that it not only eats, but wipes the plate clean in a flourish,
and leaves nothing remaining.

Starting with loading a Nul file into Memory, and then going on ahead
wreaking Havoc, NPP leaves a trail of Destruction behind, which sadly is
Irrecoverable...

**Comment by swift1978** [comment]
1) Yes, everything digital finds it's life in an OS. But a Saved File is a
Committed File, and Almost Always is left Untouched. It suffers Damage only
when the Physical Hard Disk gets corrupted, etc, or the Drive gets Locked,
Encrypted and the Password lost, or a User Blunder where he
deletes, over-writes, whatever.
Otherwise a Saved File is an Intact File transferable between OS', and
wherever, with no change in Contents. Because all the OS' that we normally
encounter don't make it their business to modify the Contents of a Saved
File; they just give it a piece of Digital Real Estate to reside on, and be
available for access.

I am pretty sure that NONE of these Errors that we are experiencing with
NPP occurred when NPP executed a Normal Save, when the System was
functioning normally.

I am not a Software/Hardware Professional, but I don't believe that during
a Crash, what happens is a Regular File Save routine, where the File on
Disk is Over-written. I don't think that happens. What if the File was an
Unsaved File residing on one of NPP's tabs.

I believe what happens is a Memory Dump, that the OS saves somewhere else
as it is coded to...

And I believe that when NPP is re-launched, it fetches the Data from the
Dumped Memory and loads it into Memory afresh. And these days, as it is the
case, what loads in Memory is a Nul file. The Files on Disk are not touched
at this Stage. Somewhere after that, depending on the Settings of the User,
the Files on Disk, and Backups are over-written by NPP or it's Plugins with
the Nul File that is in Memory at that stage. That is what I believe is the
case, after closely observing, and re-living what happened to me.

2) I am Curious about your Statement that NPP makes additional copies of
Unsaved Files in it's own Folder that have Time Stamp, and are offered for
Restore when we reload our Session.
I am using NPP v7.9 64 bit, and this is what I want set up, Timed Backups
and I am not able to do it. Wondering how you managed to set this up. I
understand that there are Backup Options.
In Mine, since forever I have kept following Options 'checked' under
'Session Snapshot and Periodic Backup'. a) Remember current Session for
next launch is always checked in my Settings. b) Enable Session Snapshot
and Periodic Backup is checked with following Settings, Backup every 7
seconds, and Backup Path, AppData\Roaming\Notepad++\Backup
I just Select it, but it NEVER WORKS for me. Not much I can do more than
that, isn't it.
The Backup Folder above never has any file. Even right now, I checked it,
and it was Empty. The Only Session file is the Current one, and that is not
in the Main NPP Folder and not the Backup Sub-Folder. That's why I am
wondering where you got your Timed Backups from. How did you manage to get
NPP to do that.
